### ISSUE TITLE
"Neumann" boundary condition

### ISSUE THREAD

**Issue by jedbrown** [barbagroup/AmgXWrapper#13]
The code in `applyNeumannBC` actually implements Dirichlet boundary conditions. This is confusing when reading the code and the problem specification. It would also be useful to state the domain and boundary conditions in the comment at the top of `main.cpp`.

**Comment by piyueh** [comment]
Yes, you're right. It's confusing.

When we create matrix A in the function `generateA`, we already apply pure Neumann BCs to the matrix A. What the function `applyNeumannBC` does is just pinning a point as a reference point (i.e. Dirichlet BC) to avoid singular matrix. So just as you said, it's not applying Neumann BC. Instead, it's applying Dirichlet BC.

I'll modify the documentation and comments and ping you once I finish them.

**Comment by jedbrown** [comment]
Fair enough. An alternative (in PETSc) is to set the constant null space, in which case everything should work without this pinning.

**Comment by piyueh** [comment]
If we use null space in PETSc, I'm not sure how to do that in AmgX. In **poisson** example, we are doing a comparison, to some degree, between PETSc and AmgX (though the main purpose is to provide an example), so we hope to keep most stuff the same between `AmgX_GPU` and `PETSc` modes.

I do agree `applyNeumannBC` is confusing. I'm working on that to make it more clear.

**Comment by piyueh** [comment]
Resolved in commit 2d2727f1528dd4ec5a15167579bc61ab4d793b87
